---
---

<center><h1> 📍 📍 Sorting the Data 📍 📍 </h1></center>

---

In this notebook, we will see how to sort the data based on a single column and on multiple columns. Also we will see how to reset index after sorting the data.

---

In [1]:
# import the pandas library
import pandas as pd

 ---

#### `CREATE A SAMPLE DATAFRAME`


- We will create a sample data of students that will contain `roll_no`, `name`, `grade`, `marks` and `city`. 


---

In [2]:
# sample dataframe
data_frame = pd.DataFrame({
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})

data_frame

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

---

#### `SORT THE DATAFRAME BY THE GRADES OF THE STUDENT` 


- Now, we need to sort the students based on their grades. We will use the **`sort_values`** function and in the parameter by we will pass the column **`grade`**. 

---

In [4]:
# sort the data
data_frame.sort_values(by=['grade'])

roll_no     name grade  marks       city
2      104  Prateek     A     20      Delhi
4      105   Kartik     A     22  Hyderabad
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
3      103  Piyuesh     C      4   Gurugram

---

- Now, that we have sorted the data based on the grades. You can see that some of the students with higher marks are ranked lower even having the same marks. So, we want to sort the data based on both grades and marks.

- Also, the grades should be in ascending order and marks should be in descending order.


---
---

#### `SORT THE DATAFRAME BY GRADE AND MARKS OF THE STUDENTS`


- To sort the values in the descending order, we need to set parameter `ascending = False`.
- In the by parameter pass the list of columns on which we want to sort and for the ascending parameter pass the boolean list `True for ascending` and `False for descending`.

---

In [5]:
# sort the dataframe
data_frame.sort_values(by=['grade','marks'],ascending=[True,False])

roll_no     name grade  marks       city
4      105   Kartik     A     22  Hyderabad
2      104  Prateek     A     20      Delhi
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
3      103  Piyuesh     C      4   Gurugram

---

***Now, let's see the dataframe.***

In [6]:
data_frame

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

***Now, you can observe that the sorted state of the dataframe is not saved. Use the parameter `inplace = True` to save the sorted state.***

---

In [7]:
# save the sorted state by using inplace =True
data_frame.sort_values(by=['grade','marks'], ascending= [True, False], inplace=True)

In [8]:
data_frame

roll_no     name grade  marks       city
4      105   Kartik     A     22  Hyderabad
2      104  Prateek     A     20      Delhi
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
3      103  Piyuesh     C      4   Gurugram

---

***Now, we can see that the index is also shuffled according to the sorting. If we want to reset the index we use `reset_index` function.***


---

In [9]:
data_frame.reset_index()

index  roll_no     name grade  marks       city
0      4      105   Kartik     A     22  Hyderabad
1      2      104  Prateek     A     20      Delhi
2      0      102  Aravind     B     15   Gurugram
3      1      101    Rahul     B     15      Delhi
4      3      103  Piyuesh     C      4   Gurugram

---

***Now, it has created another column `index` which is the previous index. If you want to remove this just pass the parameter `drop = True` and also `inplace = True` to save the state.***

---

In [10]:
data_frame.reset_index(inplace=True, drop=True)
data_frame

roll_no     name grade  marks       city
0      105   Kartik     A     22  Hyderabad
1      104  Prateek     A     20      Delhi
2      102  Aravind     B     15   Gurugram
3      101    Rahul     B     15      Delhi
4      103  Piyuesh     C      4   Gurugram

---
---

<center><h1> 📍 📍 Working with Multiple DataFrames 📍 📍 </h1></center>


---

- In most of the real life projects you will not get data from a single resource. You might need to combine data that you gather from multiple sources. In this notebook, we will see how to work with multiple data files? 


***We have already worked with big mart sales data. Here, we have divided the data based on different outlet size. Let's see how to work with the multiple files***


---

#### `READ THE DATA`

---

In [11]:
# importing the pandas library
import pandas as pd

---

***Read 3 different files `outlet_size_small.csv`, `outlet_size_medium.csv` and `outlet_size_high.csv` stored in the folder datasets.***

---

In [12]:
# read the datasets
outlet_size_small = pd.read_csv("C:\Arun\Data/outlet_size_small.csv")
outlet_size_medium = pd.read_csv("C:\Arun\Data/outlet_size_medium.csv")
outlet_size_large = pd.read_csv("C:\Arun\Data/outlet_size_high.csv")

***Let's check the shape of the data.***

---

In [13]:
outlet_size_small.shape, outlet_size_medium.shape, outlet_size_large.shape

((2388, 9), (2793, 9), (932, 9))

***So, there are 2388 small, 2793 medium and 932 large size outlets are there.***


---

***Let's have a look at the data***

#### `OUTLET SIZE SMALL`
---

In [14]:
outlet_size_small.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  
0  144.1102            OUT046  Supermarket Type1          2187.1530  
1  119.6782            OUT046  Supermarket Type1          2145.2076  
2   56.3614            OUT046  Supermarket Type1          1547.3192  
3  230.5352            OUT035  Supermarket Type1          2748.4224  
4  196.5084            OUT035  Supermarket Type1          1587.2672

#### `OUTLET SIZE MEDIUM`

---

In [15]:
outlet_size_medium.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDP36       10.395          Regular         0.000000   
4           FDP10          NaN          Low Fat         0.127470   

      Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0         Dairy  249.8092            OUT049  Supermarket Type1   
1   Soft Drinks   48.2692            OUT018  Supermarket Type2   
2          Meat  141.6180            OUT049  Supermarket Type1   
3  Baking Goods   51.4008            OUT018  Supermarket Type2   
4   Snack Foods  107.7622            OUT027  Supermarket Type3   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           556.6088  
4          4022.7636

#### `OUTLET SIZE LARGE`

---

In [16]:
outlet_size_large.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           NCD19         8.93          Low Fat         0.000000   
1           FDO10        13.65          Regular         0.012741   
2           FDF32        16.35          Low Fat         0.068024   
3           FDN22        18.85          Regular         0.138190   
4           DRJ59        11.65          low fat         0.019356   

               Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0              Household   53.8614            OUT013  Supermarket Type1   
1            Snack Foods   57.6588            OUT013  Supermarket Type1   
2  Fruits and Vegetables  196.4426            OUT013  Supermarket Type1   
3            Snack Foods  250.8724            OUT013  Supermarket Type1   
4            Hard Drinks   39.1164            OUT013  Supermarket Type1   

   Item_Outlet_Sales  
0           994.7052  
1           343.5528  
2          1977.4260  
3          3775.0860  
4           308.9312

---

#### `CONCATENATE ALL THE DATAFRAMES`


We will use the concat function to concatenate all the dataframes. You just need to pass the list of dataframes to concatenate.

---

#### `FOR ROW-WISE CONCATENATION USE AXIS=0`


---

In [17]:
# dataframes list
all_dataframes = [outlet_size_small, outlet_size_medium, outlet_size_large]

In [19]:
# concatenate all the dataframes
data = pd.concat(all_dataframes, axis=0)

In [21]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  
0  144.1102            OUT046  Supermarket Type1          2187.1530  
1  119.6782            OUT046  Supermarket Type1          2145.2076  
2   56.3614            OUT046  Supermarket Type1          1547.3192  
3  230.5352            OUT035  Supermarket Type1          2748.4224  
4  196.5084            OUT035  Supermarket Type1          1587.2672

In [20]:
# shape of the data
data.shape

(6113, 9)

#### `FOR COLUMN-WISE CONCATENATION USE AXIS=1`


It is not advised to concatenate dataframes column wise. If you want to then you need to take care of some checks like the number of rows must be same in both the dataframes. Indexes are sorted of both the dataframes. If you are done with all the checks then you can simply use `axis=1` to do the job.



Let's see with the help of an example.


---

In [22]:
sample_dataframe = pd.DataFrame({
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
sample_dataframe

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

***Let's create a dataframe which contains column name `phone_no`. Here, we are assuming that order of phone numbers are correct as the order of names in the sample_dataframe.***

---

# another sample dataframe
phone_no = pd.DataFrame({ 'phone_no' : [212202, 202021, 212334, 213431, 211721]})
phone_no

In [24]:
combined = pd.concat([sample_dataframe,phone_no], axis=1)
combined

roll_no     name grade  marks       city  phone_no
0      102  Aravind     B     15   Gurugram    212202
1      101    Rahul     B     15      Delhi    202021
2      104  Prateek     A     20      Delhi    212334
3      103  Piyuesh     C      4   Gurugram    213431
4      105   Kartik     A     22  Hyderabad    211721

---

#### `Performing SQL-Like Joins in Pandas`


#### `LET'S LEARN ABOUT THE JOINS IN PANDAS WITH SOME EXAMPLES`

- We will create a sample dataframe of students data that will contain `roll_no`, `name`, `grade`, `marks` and `city`. 


---

In [25]:
student_df = pd.DataFrame({
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

---

***Now, we have a mapping of city and state. We want to add another feature to our dataframe state using this mapping.***


Let's create the city_state_mapping 

---

In [27]:
city_state_mapping = pd.DataFrame({
    'city' :  ['Gurugram', 'Delhi', 'Hyderabad', 'Faridabad'],
    'state' : ['Haryana',  'Delhi', 'Telangana', 'Haryana']
})
city_state_mapping

city      state
0   Gurugram    Haryana
1      Delhi      Delhi
2  Hyderabad  Telangana
3  Faridabad    Haryana

***Now, we want to add another column state to the `student_df` using the `city_state_mapping`. We can do this by doing a left join. We need to use the merge function and set the parameters `how='left'` and `on='city`.***

![](left.png)

In [28]:
city_state_mapping

city      state
0   Gurugram    Haryana
1      Delhi      Delhi
2  Hyderabad  Telangana
3  Faridabad    Haryana

In [29]:
student_df.merge(city_state_mapping, how='left', on='city')

roll_no     name grade  marks       city      state
0      102  Aravind     B     15   Gurugram    Haryana
1      101    Rahul     B     15      Delhi      Delhi
2      104  Prateek     A     20      Delhi      Delhi
3      103  Piyuesh     C      4   Gurugram    Haryana
4      105   Kartik     A     22  Hyderabad  Telangana

***Now, we have another dataframe that contains roll_no of some students. We need to find out the other details of the students. We can do this by using `right join`. You just need to set `how='right'`.***

![](right.png)
---

In [30]:
roll_no = pd.DataFrame({
    'roll_no' : [ 102, 103]
})
roll_no

roll_no
0      102
1      103

In [31]:
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

In [32]:
student_df.merge(roll_no, how='right', on='roll_no')

roll_no     name grade  marks      city
0      102  Aravind     B     15  Gurugram
1      103  Piyuesh     C      4  Gurugram

***Now, students participated in a placement drive and some of the students got placed and the details are given in the `student_selection` dataframe.***

---

In [33]:
student_selection = pd.DataFrame({
    'roll_no' : [102, 105, 101],
    'company' : ['ABC', 'XYZ', 'ABC'],
    'package (lpa)' : [ 8, 14.5, 11 ]
})
student_selection

roll_no company  package (lpa)
0      102     ABC            8.0
1      105     XYZ           14.5
2      101     ABC           11.0

***Now, we want to combine the `student_df` and `student_selection`. We can do this by using `outer/full join`. You need to set parameter `how = 'outer'`.***

![](outer.png)

---

In [34]:
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

In [35]:
student_df.merge(student_selection, how='outer')

roll_no     name grade  marks       city company  package (lpa)
0      101    Rahul     B     15      Delhi     ABC           11.0
1      102  Aravind     B     15   Gurugram     ABC            8.0
2      103  Piyuesh     C      4   Gurugram     NaN            NaN
3      104  Prateek     A     20      Delhi     NaN            NaN
4      105   Kartik     A     22  Hyderabad     XYZ           14.5

---

***Now, consider it was a pool placement drive and students from multiple colleges participated. The college `"ZU UNIVERSITY"` got the list of students selected for the job. You need to find out the details of the students who got selected from the college `"ZU UNIVERSITY"`***


---

In [36]:
student_df = pd.DataFrame({
    'college': ['ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY'],
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
student_df

college  roll_no     name grade  marks       city
0  ZU UNIVERSITY      102  Aravind     B     15   Gurugram
1  ZU UNIVERSITY      101    Rahul     B     15      Delhi
2  ZU UNIVERSITY      104  Prateek     A     20      Delhi
3  ZU UNIVERSITY      103  Piyuesh     C      4   Gurugram
4  ZU UNIVERSITY      105   Kartik     A     22  Hyderabad

In [37]:
pool = pd.DataFrame({
        'college': ['ZU UNIVERSITY', 'ZU UNIVERSITY', 'AB UNIVERSITY', 'ZU UNIVERSITY','AB UNIVERSITY'],
        'name' : ['Aravind', 'Rahul', 'Rahul', 'Prateek', 'Harsh'],
        'company' : ['ABC', 'XYZ', 'ABC', 'AEP', 'ABC'],
        'package (lpa)' : [ 8, 14.5, 11, 6, 6 ]
})
pool

college     name company  package (lpa)
0  ZU UNIVERSITY  Aravind     ABC            8.0
1  ZU UNIVERSITY    Rahul     XYZ           14.5
2  AB UNIVERSITY    Rahul     ABC           11.0
3  ZU UNIVERSITY  Prateek     AEP            6.0
4  AB UNIVERSITY    Harsh     ABC            6.0

---

***Now, we have 2 columns common `college` and `name` in both the dataframes. So, here we will use the `inner join`. You just need to set the parameters `how='inner'` and as we have 2 common columns therefore, set the parameter `on=['college','name']`.***


![](inner.png)

---

In [39]:
student_df.merge(pool, how='inner', on=['college', 'name'])

college  roll_no     name grade  marks      city company  \
0  ZU UNIVERSITY      102  Aravind     B     15  Gurugram     ABC   
1  ZU UNIVERSITY      101    Rahul     B     15     Delhi     XYZ   
2  ZU UNIVERSITY      104  Prateek     A     20     Delhi     AEP   

   package (lpa)  
0            8.0  
1           14.5  
2            6.0

---

### `NOW LET'S SOLVE THE PROBLEM`

---

We have another dataset `outlet_data.csv` in the dataset folder. It has column `Outlet_Identifier`, `Outlet_Establishment_Year`, `Outlet_Size` and `Outlet_Location_Type`.


Now, we have `Outlet_Identifier` in both the datasets and we need to combine them and get the rest of the variables in our dataset. So we will do a `left join` to merge the data frames. Let's see how?

---


#### `READ THE OUTLET DATA`

---

In [40]:
### all merges with sample 

In [41]:
previous_data = pd.read_csv("C:\Arun\Data/outlet_size_concatenated_data.csv")

In [42]:
previous_data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  
0  144.1102            OUT046  Supermarket Type1          2187.1530  
1  119.6782            OUT046  Supermarket Type1          2145.2076  
2   56.3614            OUT046  Supermarket Type1          1547.3192  
3  230.5352            OUT035  Supermarket Type1          2748.4224  
4  196.5084            OUT035  Supermarket Type1          1587.2672

In [43]:
# read the outlet data
outlet_data = pd.read_csv("C:\Arun\Data/outlet_data.csv")

In [44]:
# view the top rows of the data
outlet_data.head()

Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0            OUT013                       1987        High   
1            OUT018                       2009      Medium   
2            OUT019                       1985       Small   
3            OUT027                       1985      Medium   
4            OUT035                       2004       Small   

  Outlet_Location_Type  
0               Tier 3  
1               Tier 3  
2               Tier 1  
3               Tier 3  
4               Tier 2

---


***Use the merge function to set parameter `how = 'left'` for the left join and the set the on parameter as the common column name as `on='Outlet_Identifier'`.***


---

In [45]:
# merge the data
combined_data = previous_data.merge(outlet_data, how='left', on='Outlet_Identifier')

In [46]:
# view the data
combined_data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  \
0  144.1102            OUT046  Supermarket Type1          2187.1530   
1  119.6782            OUT046  Supermarket Type1          2145.2076   
2   56.3614            OUT046  Supermarket Type1          1547.3192   
3  230.5352            OUT035  Supermarket Type1          2748.4224   
4  196.5084            OUT035  Supermarket Type1          1587.2672   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  
0                       1997       Small               Tier 1  
1                       1997       Small               Tier 1  
2                       1997       Small               Tier 1  
3                       2004       Small               Tier 2  
4                       2004       Small               Tier 2

In [47]:
combined_data.shape

(6113, 12)

----

So, you can see that we got some extra variables in the dataset.

----


***Now, we have a sample list of Item_Identifiers in a separate file `item_identifier.csv`. We are required to provide all the details that we have related to that particular Item_Identifier. Let's see how can we do that with the help of a `RIGHT JOIN`***

---


#### `READ THE DATA`

---

In [48]:
# read the data
item_data = pd.read_csv("C:\Arun\Data/item_idenifier.csv")

In [49]:
# view the top rows
item_data.head()

Item_Identifier
0           DRI51
1           FDL48
2           FDL38
3           FDF17
4           FDN56

In [50]:
item_data.shape

(100, 1)

***So, we have 100 Item_Identifiers and we need to provide the other details. We just need to use the merge function and set parameter `how='right'` and the common variable name is `Item_Identifier`.***

---

In [51]:
item_details = combined_data.merge(item_data, how='right',on='Item_Identifier')

In [52]:
item_details

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0             DRI51        17.25          Low Fat         0.042234   
1             DRI51        17.25          Low Fat         0.042414   
2             DRI51          NaN          Low Fat         0.042037   
3             FDL48        19.35          Regular         0.082251   
4             FDL48        19.35          Regular         0.082266   
..              ...          ...              ...              ...   
417           FDD57        18.10               LF         0.022381   
418           FDG41         8.84          Regular         0.076548   
419           FDG41         8.84          Regular         0.076681   
420           FDG41         8.84          Regular         0.076874   
421           NCM42         6.13          Low Fat         0.028366   

                 Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0                    Dairy  173.3764            OUT035  Supermarket Type1   
1                    Dairy  173.1764            OUT018  Supermarket Type2   
2                    Dairy  172.6764            OUT027  Supermarket Type3   
3             Baking Goods   48.7034            OUT035  Supermarket Type1   
4             Baking Goods   48.8034            OUT046  Supermarket Type1   
..                     ...       ...               ...                ...   
417  Fruits and Vegetables   93.6094            OUT013  Supermarket Type1   
418           Frozen Foods  109.5228            OUT035  Supermarket Type1   
419           Frozen Foods  110.7228            OUT049  Supermarket Type1   
420           Frozen Foods  109.9228            OUT018  Supermarket Type2   
421              Household  110.0912            OUT049  Supermarket Type1   

     Item_Outlet_Sales  Outlet_Establishment_Year Outlet_Size  \
0            2061.3168                       2004       Small   
1            4466.1864                       2009      Medium   
2            6183.9504                       1985      Medium   
3             534.6374                       2004       Small   
4             340.2238                       1997       Small   
..                 ...                        ...         ...   
417           476.0470                       1987        High   
418          1657.8420                       2004       Small   
419          1657.8420                       1999      Medium   
420          1547.3192                       2009      Medium   
421          1637.8680                       1999      Medium   

    Outlet_Location_Type  
0                 Tier 2  
1                 Tier 3  
2                 Tier 3  
3                 Tier 2  
4                 Tier 1  
..                   ...  
417               Tier 3  
418               Tier 2  
419               Tier 1  
420               Tier 3  
421               Tier 1  

[422 rows x 12 columns]

---

<center><h1> 📍 📍 Aggregating and Summarizing DataFrames 📍 📍 </h1></center>

---

- How to calculate sum, mean, median and mode of a column?
- How to get the summary of the numerical variables?
- How to get number of missing values in each columns?
- How to group the data based on categories of one column?
- How to group the data based on categories of multiple columns?
- How to create new feature using the aggregated results of a column?

---

#### `READ THE DATASET`

- In this notebook, we are going to use the big mart sales data that we have used previously. It is stored in the folder name `datasets`.

In [53]:
# import the pandas library
import pandas as pd

In [54]:
# read the big mart sales data
data = pd.read_csv("C:\Arun\Data/big_mart_sales.csv")

In [55]:
# view the top rows of the data
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

---

#### `How to calculate sum, mean, median and mode of a column?`

---

In [59]:
# calculate the sum of all MRP
data.Item_MRP.sum()

np.float64(1201681.4808)

In [60]:
# calculate the average Item_MRP
data.Item_MRP.mean()

np.float64(140.9927819781767)

In [61]:
# calculate the median of Item_MRP
data.Item_MRP.median()

143.0128

In [62]:
# find out the most frequent outlet type
data.Outlet_Type.mode()

0    Supermarket Type1
Name: Outlet_Type, dtype: object

In [63]:
data.Outlet_Type.value_counts()

Outlet_Type
Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: count, dtype: int64

---

#### `How to get the summary of the numerical variables?`

To get the summary of the numerical variables we have describe function in pandas.

---

In [64]:
data.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

---

#### `How to get number of missing values in each columns?`

---

In [65]:
data.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

----
#### `How to group the data based on categories of one column?`
---

#### `GROUP BY`


- Calculate the average MRP of each `Item_Type` using groupby.

---

In [66]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [67]:
d = data.groupby(['Item_Type'])
d

In [70]:
d.first()

Item_Identifier  Item_Weight Item_Fat_Content  \
Item_Type                                                             
Baking Goods                    FDP36       10.395          Regular   
Breads                          FDO23       17.850          Low Fat   
Breakfast                       FDP49        9.000          Regular   
Canned                          FDC14       21.350          Regular   
Dairy                           FDA15        9.300          Low Fat   
Frozen Foods                    FDH17       16.200          Regular   
Fruits and Vegetables           FDX07       19.200          Regular   
Hard Drinks                     DRI11       11.650          Low Fat   
Health and Hygiene              NCB42       11.800          Low Fat   
Household                       NCD19        8.930          Low Fat   
Meat                            FDN15       17.500          Low Fat   
Others                          NCN07       18.500          Low Fat   
Seafood                         FDG33       10.395          Regular   
Snack Foods                     FDO10       13.650          Regular   
Soft Drinks                     DRC01        5.920          Regular   
Starchy Foods                   FDH35       18.250          Low Fat   

                       Item_Visibility  Item_MRP Outlet_Identifier  \
Item_Type                                                            
Baking Goods                  0.000000   51.4008            OUT018   
Breads                        0.000000   93.1436            OUT045   
Breakfast                     0.069089   56.3614            OUT046   
Canned                        0.072222   43.6454            OUT019   
Dairy                         0.016047  249.8092            OUT049   
Frozen Foods                  0.016687   96.9726            OUT045   
Fruits and Vegetables         0.000000  182.0950            OUT010   
Hard Drinks                   0.034238  113.2834            OUT027   
Health and Hygiene            0.008596  115.3492            OUT018   
Household                     0.000000   53.8614            OUT013   
Meat                          0.016760  141.6180            OUT049   
Others                        0.056816  132.1284            OUT010   
Seafood                       0.139561  170.4764            OUT027   
Snack Foods                   0.012741   57.6588            OUT013   
Soft Drinks                   0.019278   48.2692            OUT018   
Starchy Foods                 0.000000  164.7526            OUT045   

                       Outlet_Establishment_Year Outlet_Size  \
Item_Type                                                      
Baking Goods                                2009      Medium   
Breads                                      2002      Medium   
Breakfast                                   1997       Small   
Canned                                      1985       Small   
Dairy                                       1999      Medium   
Frozen Foods                                2002       Small   
Fruits and Vegetables                       1998      Medium   
Hard Drinks                                 1985      Medium   
Health and Hygiene                          2009      Medium   
Household                                   1987        High   
Meat                                        1999      Medium   
Others                                      1998       Small   
Seafood                                     1985      Medium   
Snack Foods                                 1987        High   
Soft Drinks                                 2009      Medium   
Starchy Foods                               2002      Medium   

                      Outlet_Location_Type        Outlet_Type  \
Item_Type                                                       
Baking Goods                        Tier 3  Supermarket Type2   
Breads                              Tier 2  Supermarket Type1   
Breakfast                           Tier 1  Supermarket Type1   
Cann

#### `GROUPBY MEAN`

---

In [73]:
d.mean(numeric_only=True)

Item_Weight  Item_Visibility    Item_MRP  \
Item_Type                                                         
Baking Goods             12.277108         0.069169  126.380766   
Breads                   11.346936         0.066255  140.952669   
Breakfast                12.768202         0.085723  141.788151   
Canned                   12.305705         0.068129  139.763832   
Dairy                    13.426069         0.072427  148.499208   
Frozen Foods             12.867061         0.065645  138.503366   
Fruits and Vegetables    13.224769         0.068513  144.581235   
Hard Drinks              11.400328         0.064943  137.077928   
Health and Hygiene       13.142314         0.055216  130.818921   
Household                13.384736         0.061322  149.424753   
Meat                     12.817344         0.062284  139.882032   
Others                   13.853285         0.060241  132.851430   
Seafood                  12.552843         0.074976  141.841719   
Snack Foods              12.987880         0.066850  146.194934   
Soft Drinks              11.847460         0.063972  131.492506   
Starchy Foods            13.690731         0.067564  147.838023   

                       Outlet_Establishment_Year  Item_Outlet_Sales  
Item_Type                                                            
Baking Goods                         1997.728395        1952.971207  
Breads                               1997.657371        2204.132226  
Breakfast                            1997.336364        2111.808651  
Canned                               1998.152542        2225.194904  
Dairy                                1997.681818        2232.542597  
Frozen Foods                         1998.024533        2132.867744  
Fruits and Vegetables                1997.719968        2289.009592  
Hard Drinks                          1998.177570        2139.221622  
Health and Hygiene                   1997.734615        2010.000265  
Household                            1997.784615        2258.784300  
Meat                                 1997.341176        2158.977911  
Others                               1997.662722        1926.139702  
Seafood                              1997.734375        2326.065928  
Snack Foods                          1997.886667        2277.321739  
Soft Drinks                          1998.098876        2006.511735  
Starchy Foods                        1998.581081        2374.332773

In [75]:
d['Item_MRP'].mean()

Item_Type
Baking Goods             126.380766
Breads                   140.952669
Breakfast                141.788151
Canned                   139.763832
Dairy                    148.499208
Frozen Foods             138.503366
Fruits and Vegetables    144.581235
Hard Drinks              137.077928
Health and Hygiene       130.818921
Household                149.424753
Meat                     139.882032
Others                   132.851430
Seafood                  141.841719
Snack Foods              146.194934
Soft Drinks              131.492506
Starchy Foods            147.838023
Name: Item_MRP, dtype: float64

#### `GROUPBY MAX`

---

In [77]:
d.max(numeric_only=True)

Item_Weight  Item_Visibility  Item_MRP  \
Item_Type                                                       
Baking Goods                 20.85         0.311090  265.5568   
Breads                       20.85         0.281510  263.6594   
Breakfast                    21.10         0.274592  234.9300   
Canned                       21.35         0.328391  266.8884   
Dairy                        20.70         0.304737  266.6884   
Frozen Foods                 20.85         0.294939  264.8910   
Fruits and Vegetables        21.35         0.321115  264.2252   
Hard Drinks                  19.70         0.298205  261.4278   
Health and Hygiene           21.25         0.255348  266.6884   
Household                    21.25         0.325781  264.7910   
Meat                         21.25         0.287017  261.5936   
Others                       20.50         0.210596  254.9040   
Seafood                      20.75         0.256152  240.2880   
Snack Foods                  21.25         0.303743  264.9568   
Soft Drinks                  20.75         0.306543  263.5910   
Starchy Foods                21.20         0.238366  263.0252   

                       Outlet_Establishment_Year  Item_Outlet_Sales  
Item_Type                                                            
Baking Goods                                2009          7931.6754  
Breads                                      2009          8958.3390  
Breakfast                                   2009          8209.3140  
Canned                                      2009         10306.5840  
Dairy                                       2009         10256.6490  
Frozen Foods                                2009          9678.0688  
Fruits and Vegetables                       2009         12117.5600  
Hard Drinks                                 2009          7843.1240  
Health and Hygiene                          2009          9779.9362  
Household                                   2009         13086.9648  
Meat                                        2009          9390.4432  
Others                                      2009          6008.8450  
Seafood                                     2009          6503.5344  
Snack Foods                                 2009         10993.6896  
Soft Drinks                                 2009          9554.2300  
Starchy Foods                               2009          8132.0812

---

#### `How to group the data based on categories of multiple columns?`


- Calculate the `average MRP` of each `Item_Type` for each category of `Outlet_Size`.


---

In [79]:
# Group by Outlet_Size and Item_Type, compute mean for only numeric columns
grouped_data = data.groupby(['Outlet_Size', 'Item_Type']).mean(numeric_only=True)
print(grouped_data)


                                   Item_Weight  Item_Visibility    Item_MRP  \
Outlet_Size Item_Type                                                         
High        Baking Goods             12.036918         0.059490  129.202044   
            Breads                   11.048000         0.065586  133.758960   
            Breakfast                12.564231         0.075118  147.490585   
            Canned                   11.922231         0.056733  135.442708   
            Dairy                    13.071875         0.068907  153.509173   
            Frozen Foods             13.250707         0.065639  136.829250   
            Fruits and Vegetables    13.259613         0.061302  145.572870   
            Hard Drinks              11.741957         0.062271  141.927522   
            Health and Hygiene       13.028770         0.051031  135.110980   
            Household                14.033398         0.053742  147.097522   
            Meat                     13.384268      

In [85]:
d = data.groupby(['Outlet_Size', 'Item_Type']).mean(numeric_only=True)

In [86]:
d

Item_Weight  Item_Visibility    Item_MRP  \
Outlet_Size Item_Type                                                         
High        Baking Goods             12.036918         0.059490  129.202044   
            Breads                   11.048000         0.065586  133.758960   
            Breakfast                12.564231         0.075118  147.490585   
            Canned                   11.922231         0.056733  135.442708   
            Dairy                    13.071875         0.068907  153.509173   
            Frozen Foods             13.250707         0.065639  136.829250   
            Fruits and Vegetables    13.259613         0.061302  145.572870   
            Hard Drinks              11.741957         0.062271  141.927522   
            Health and Hygiene       13.028770         0.051031  135.110980   
            Household                14.033398         0.053742  147.097522   
            Meat                     13.384268         0.055831  137.244790   
            Others                   14.008437         0.049660  132.576613   
            Seafood                  13.002000         0.066497  134.864240   
            Snack Foods              13.404560         0.059669  145.847086   
            Soft Drinks              12.233776         0.053552  131.758473   
            Starchy Foods            13.562632         0.073566  158.157074   
Medium      Baking Goods             12.170112         0.060220  126.178568   
            Breads                   11.679231         0.057480  140.861039   
            Breakfast                11.668800         0.071658  134.537511   
            Canned                   12.427517         0.063452  138.124851   
            Dairy                    13.840695         0.063799  148.512174   
            Frozen Foods             12.619946         0.059927  140.557015   
            Fruits and Vegetables    13.212875         0.060720  142.971470   
            Hard Drinks              11.160288         0.058881  142.837696   
            Health and Hygiene       13.579727         0.048143  128.701865   
            Household                13.683237         0.056420  147.711330   
            Meat                     12.372527         0.064558  136.419132   
            Others                   13.850811         0.057396  127.836181   
            Seafood                  14.322500         0.081870  140.857619   
            Snack Foods              12.908487         0.061292  148.779195   
            Soft Drinks              11.679565         0.060093  128.269682   
            Starchy Foods            14.289265         0.057441  148.419504   
Small       Baking Goods             12.644236         0.073599  125.213364   
            Breads                   10.792000         0.072865  145.523651   
            Breakfast                13.930500         0.098983  130.568027   
            Canned                   12.351060         0.071590  142.295429   
            Dairy                    13.395503         0.081580  145.942101   
            Frozen Foods             12.890000         0.069465  137.838544   
            Fruits and Vegetables    13.144647         0.075157  148.313370   
            Hard Drinks              11.646190         0.062429  129.758784   
            Health and Hygiene       13.100330         0.059809  131.831535   
            Household                13.388537         0.066570  153.965439   
            Meat                     12.836839         0.062318  145.699250   
            Others                   13.541316         0.058949  137.889211   
            Seafood                  12.527500         0.085069  144.281760   
            Snack Foods              12.859615         0.074044  144.351896   
            Soft Drinks              11.906300         0.070324  132.855043   
            Starchy Foods            13.245294         0.066771  150.270174   

                                   Outlet_Establishment_Year  \
Outlet_Size Item_Typ

---

#### `We can also do this using the pivot table also`


- Learn more about the pivot table here: 
    - https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html
    - https://www.analyticsvidhya.com/blog/2020/03/pivot-table-pandas-python/

---
Calculate the average MRP of each `Item_Type` using pivot table.

---

In [87]:
pd.pivot_table(data, index='Item_Type', values="Item_MRP", aggfunc= 'mean')

Item_MRP
Item_Type                        
Baking Goods           126.380766
Breads                 140.952669
Breakfast              141.788151
Canned                 139.763832
Dairy                  148.499208
Frozen Foods           138.503366
Fruits and Vegetables  144.581235
Hard Drinks            137.077928
Health and Hygiene     130.818921
Household              149.424753
Meat                   139.882032
Others                 132.851430
Seafood                141.841719
Snack Foods            146.194934
Soft Drinks            131.492506
Starchy Foods          147.838023

---

- Calculate the `average MRP` of each `Item_Type` for each category of `Outlet_Size` using pivot table

---

In [88]:
pd.pivot_table(data, index= ['Outlet_Size', 'Item_Type'], values= 'Item_MRP', aggfunc= 'mean')

Item_MRP
Outlet_Size Item_Type                        
High        Baking Goods           129.202044
            Breads                 133.758960
            Breakfast              147.490585
            Canned                 135.442708
            Dairy                  153.509173
            Frozen Foods           136.829250
            Fruits and Vegetables  145.572870
            Hard Drinks            141.927522
            Health and Hygiene     135.110980
            Household              147.097522
            Meat                   137.244790
            Others                 132.576613
            Seafood                134.864240
            Snack Foods            145.847086
            Soft Drinks            131.758473
            Starchy Foods          158.157074
Medium      Baking Goods           126.178568
            Breads                 140.861039
            Breakfast              134.537511
            Canned                 138.124851
            Dairy                  148.512174
            Frozen Foods           140.557015
            Fruits and Vegetables  142.971470
            Hard Drinks            142.837696
            Health and Hygiene     128.701865
            Household              147.711330
            Meat                   136.419132
            Others                 127.836181
            Seafood                140.857619
            Snack Foods            148.779195
            Soft Drinks            128.269682
            Starchy Foods          148.419504
Small       Baking Goods           125.213364
            Breads                 145.523651
            Breakfast              130.568027
            Canned                 142.295429
            Dairy                  145.942101
            Frozen Foods           137.838544
            Fruits and Vegetables  148.313370
            Hard Drinks            129.758784
            Health and Hygiene     131.831535
            Household              153.965439
            Meat                   145.699250
            Others                 137.889211
            Seafood                144.281760
            Snack Foods            144.351896
            Soft Drinks            132.855043
            Starchy Foods          150.270174

---

#### `CROSS TAB`


- The crosstab() function is used to compute a frequency table of two or more factors. 
- By default, it computes a frequency table of the factors unless an array of values or an aggregation function which is passed. 

- Learn more about crosstab here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html


---

In [89]:
pd.crosstab(data['Outlet_Size'], data['Item_Type'])

Item_Type    Baking Goods  Breads  Breakfast  Canned  Dairy  Frozen Foods  \
Outlet_Size                                                                 
High                   73      25         13      65     80            92   
Medium                203      83         36     217    218           274   
Small                 187      71         30     189    198           249   

Item_Type    Fruits and Vegetables  Hard Drinks  Health and Hygiene  \
Outlet_Size                                                           
High                           142           23                  61   
Medium                         413           75                 170   
Small                          328           50                 136   

Item_Type    Household  Meat  Others  Seafood  Snack Foods  Soft Drinks  \
Outlet_Size                                                               
High               103    41      16        5          125           49   
Medium             289   149      52       21          408          137   
Small              257   119      55       20          335          126   

Item_Type    Starchy Foods  
Outlet_Size                 
High                    19  
Medium                  48  
Small                   38

---

#### `How to create new feature using the aggregated results of a column?`

Let's create a dataframe which contains Item_Identifier and the average Item_Visibility using the groupby function. 

---

In [90]:
average_item_visibility = data.groupby(['Item_Identifier'])['Item_Visibility'].mean().reset_index()
average_item_visibility

Item_Identifier  Item_Visibility
0              DRA12         0.031956
1              DRA24         0.048062
2              DRA59         0.134718
3              DRB01         0.082126
4              DRB13         0.008002
...              ...              ...
1554           NCZ30         0.024956
1555           NCZ41         0.051623
1556           NCZ42         0.009044
1557           NCZ53         0.027775
1558           NCZ54         0.080426

[1559 rows x 2 columns]

---

#### Now, we want to create a new feature `average_item_visibility` using the above dataframe. 

Let's first define a function that will take the parameter `Item_Identifier` and return the corresponding average `Item_Visibilty` using the dataframe `average_item_visibility`

---

In [91]:
def get_item_visibility(x) :
    return average_item_visibility.loc[(average_item_visibility.Item_Identifier == x), 'Item_Visibility'].values[0]

In [92]:
# let's test it on the sample Item_Identifier
get_item_visibility('DRA24')

np.float64(0.04806226414285714)

---

Now, use the apply function to create the new feature. You just need to access the Item_Identifier column and use the apply method and pass the function that we have defined.

---

In [93]:
data['average_item_visibility'] = data.Item_Identifier.apply(get_item_visibility)

In [94]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  average_item_visibility  
0  Supermarket Type1          3735.1380                 0.017387  
1  Supermarket Type2           443.4228                 0.019219  
2  Supermarket Type1          2097.2700                 0.020145  
3      Grocery Store           732.3800                 0.015274  
4  Supermarket Type1           994.7052                 0.008082

####  There is a better way of doing the above task is by using the transform function.

- The time taken by the transform function to perform the above operation is comparatively less over a large dataframe. That’s a sigificant advantage as comapred to the first approach we used.
- You can learn more about the transform function here: https://www.analyticsvidhya.com/blog/2020/03/understanding-transform-function-python/


---

In [95]:
data['average_item_visibility_2'] = data.groupby(['Item_Identifier'])['Item_Visibility'].transform('mean')

In [96]:
data[['average_item_visibility', 'average_item_visibility_2']]

average_item_visibility  average_item_visibility_2
0                    0.017387                   0.017387
1                    0.019219                   0.019219
2                    0.020145                   0.020145
3                    0.015274                   0.015274
4                    0.008082                   0.008082
...                       ...                        ...
8518                 0.061705                   0.061705
8519                 0.046952                   0.046952
8520                 0.035203                   0.035203
8521                 0.120686                   0.120686
8522                 0.051565                   0.051565

[8523 rows x 2 columns]

---

<center><h1>📍 📍 Preprocessing Timeseries Data 📍 📍</h1></center>

---


**`Pandas`** contains extensive capabilities and features for working with time series data for all domains. Using the NumPy datetime64 and timedelta64 dtypes, pandas has consolidated a large number of features from other Python libraries like scikits.timeseries as well as created a tremendous amount of new functionality for manipulating time series data.

---

In [97]:
# importing the pandas library and datetime module of Python Standard Library
import pandas as pd
import datetime

In [98]:
# read the data set
data = pd.read_csv("C:\Arun\Data/time_series.csv")

In [99]:
data.head()

ID          Datetime  Count
0   0  25-08-2012 00:00      8
1   1  25-08-2012 01:00      2
2   2  25-08-2012 02:00      6
3   3  25-08-2012 03:00      2
4   4  25-08-2012 04:00      2

---

***Data Types of columns***

---

In [101]:
data.dtypes

ID           int64
Datetime    object
Count        int64
dtype: object

***By Default: All datetime based columns are considered as strings.***

---

#### `CHANGE THE DATA TYPE TO DATETIME`

---

In [102]:
# change type to datetime
data.Datetime = pd.to_datetime(data.Datetime)

C:\Users\dell\AppData\Local\Temp\ipykernel_7416\2161098256.py:2: UserWarning: Parsing dates in %d-%m-%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data.Datetime = pd.to_datetime(data.Datetime)


In [103]:
# check the data types again
data.dtypes

ID                   int64
Datetime    datetime64[ns]
Count                int64
dtype: object

In [104]:
data

ID            Datetime  Count
0          0 2012-08-25 00:00:00      8
1          1 2012-08-25 01:00:00      2
2          2 2012-08-25 02:00:00      6
3          3 2012-08-25 03:00:00      2
4          4 2012-08-25 04:00:00      2
...      ...                 ...    ...
18283  18283 2014-09-25 19:00:00    868
18284  18284 2014-09-25 20:00:00    732
18285  18285 2014-09-25 21:00:00    702
18286  18286 2014-09-25 22:00:00    580
18287  18287 2014-09-25 23:00:00    534

[18288 rows x 3 columns]

In [105]:
data.Datetime.apply(lambda x: x.day_name())

0        Saturday
1        Saturday
2        Saturday
3        Saturday
4        Saturday
           ...   
18283    Thursday
18284    Thursday
18285    Thursday
18286    Thursday
18287    Thursday
Name: Datetime, Length: 18288, dtype: object

In [106]:
data.Datetime.apply(lambda x: x.month_name())

0           August
1           August
2           August
3           August
4           August
           ...    
18283    September
18284    September
18285    September
18286    September
18287    September
Name: Datetime, Length: 18288, dtype: object

---

***Let's see another data file with the time series data.***

---

In [107]:
data_2 = pd.read_csv("C:\Arun\Data/time_series_2.csv")

In [108]:
data_2

ID      Datetime  Count
0          0  25 Aug 2012       8
1          1  25 Aug 2012       2
2          2  25 Aug 2012       6
3          3  25 Aug 2012       2
4          4  25 Aug 2012       2
...      ...           ...    ...
18283  18283  25 Sep 2014     868
18284  18284  25 Sep 2014     732
18285  18285  25 Sep 2014     702
18286  18286  25 Sep 2014     580
18287  18287  25 Sep 2014     534

[18288 rows x 3 columns]

---

***We can read the date time by specifying the format.***

***Here are some of the common used directives.***

----
| **Directive** | **Meaning**                                            |
| ---           | ---                                                    |
|  **%a**       | Weekday as locale’s abbreviated name.                  |
|  **%A**       | Weekday as locale’s full name.                         |  
|  **%d**       | Day of the month as a zero-padded decimal number.      |
|  **%b**       | Month as locale’s abbreviated name.	                 |
|  **%B**       | Month as locale’s full name.	                         |
|  **%m**       | Month as a zero-padded decimal number.                 |
|  **%y**       | Year without century as a zero-padded decimal number.  |
|  **%Y**       | Year with century as a decimal number.                 |
|  **%H**       | Hour (24-hour clock) as a zero-padded decimal number.  |

---

***You can read more about the other directives and datetime library here: https://docs.python.org/3/library/datetime.html***

---

In [109]:
# convert to datetime by specifying the data time format
data_2.Datetime = pd.to_datetime(data_2.Datetime, format="%d %b %Y ")

In [110]:
data_2.head()

ID   Datetime  Count
0   0 2012-08-25      8
1   1 2012-08-25      2
2   2 2012-08-25      6
3   3 2012-08-25      2
4   4 2012-08-25      2

---

#### `TIME BASED FEATURES`

---


---

***Create features like `month` and `month_name` from the data.***

---

In [111]:
data

ID            Datetime  Count
0          0 2012-08-25 00:00:00      8
1          1 2012-08-25 01:00:00      2
2          2 2012-08-25 02:00:00      6
3          3 2012-08-25 03:00:00      2
4          4 2012-08-25 04:00:00      2
...      ...                 ...    ...
18283  18283 2014-09-25 19:00:00    868
18284  18284 2014-09-25 20:00:00    732
18285  18285 2014-09-25 21:00:00    702
18286  18286 2014-09-25 22:00:00    580
18287  18287 2014-09-25 23:00:00    534

[18288 rows x 3 columns]

In [112]:
# create month and month_name 
data['month'] = data.Datetime.dt.month
data['month_name'] = data.Datetime.dt.month_name()

In [113]:
data

ID            Datetime  Count  month month_name
0          0 2012-08-25 00:00:00      8      8     August
1          1 2012-08-25 01:00:00      2      8     August
2          2 2012-08-25 02:00:00      6      8     August
3          3 2012-08-25 03:00:00      2      8     August
4          4 2012-08-25 04:00:00      2      8     August
...      ...                 ...    ...    ...        ...
18283  18283 2014-09-25 19:00:00    868      9  September
18284  18284 2014-09-25 20:00:00    732      9  September
18285  18285 2014-09-25 21:00:00    702      9  September
18286  18286 2014-09-25 22:00:00    580      9  September
18287  18287 2014-09-25 23:00:00    534      9  September

[18288 rows x 5 columns]

---

***Create `day_name`, `day_of_week`, & `day_of_year`***

---

In [114]:
# create features
data['day_name'] = data.Datetime.dt.day_name()
data['day_of_week'] = data.Datetime.dt.dayofweek
data['day_of_year'] = data.Datetime.dt.dayofyear

In [115]:
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year  
0          238  
1          238  
2          238  
3          238  
4          238


---

#### `DIFFERENCE BETWEEN 2 DATES`

---


***Add the current date in the new column***

---

In [116]:
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year  
0          238  
1          238  
2          238  
3          238  
4          238

In [117]:
data['today'] = pd.to_datetime(datetime.date.today())

In [119]:
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year      today  
0          238 2025-09-02  
1          238 2025-09-02  
2          238 2025-09-02  
3          238 2025-09-02  
4          238 2025-09-02

In [120]:
difference_of_dates = data['today'] - data['Datetime']

In [121]:
difference_of_dates

0       4756 days 00:00:00
1       4755 days 23:00:00
2       4755 days 22:00:00
3       4755 days 21:00:00
4       4755 days 20:00:00
               ...        
18283   3994 days 05:00:00
18284   3994 days 04:00:00
18285   3994 days 03:00:00
18286   3994 days 02:00:00
18287   3994 days 01:00:00
Length: 18288, dtype: timedelta64[ns]

In [122]:
difference_of_dates.apply(lambda x: x.days)

0        4756
1        4755
2        4755
3        4755
4        4755
         ... 
18283    3994
18284    3994
18285    3994
18286    3994
18287    3994
Length: 18288, dtype: int64

In [123]:
data['day_difference'] = difference_of_dates.apply(lambda x: x.days)

In [124]:
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year      today  day_difference  
0          238 2025-09-02            4756  
1          238 2025-09-02            4755  
2          238 2025-09-02            4755  
3          238 2025-09-02            4755  
4          238 2025-09-02            4755

---

### `CHALLENGES WITH TIME DATA`

---

#### DEALING WITH TIME ZONES
 
- If you have the dataset of a specific time zone. You can tell pandas about the local time zone and later you can convert it into different time zones.
- Use function `dt.tz_localize` to set the local time zone.
 
 
---

In [125]:
# set the current time as of Asia
data['asia_timezone'] = data.Datetime.dt.tz_localize('Asia/Calcutta')

In [126]:
# view the data
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year      today  day_difference             asia_timezone  
0          238 2025-09-02            4756 2012-08-25 00:00:00+05:30  
1          238 2025-09-02            4755 2012-08-25 01:00:00+05:30  
2          238 2025-09-02            4755 2012-08-25 02:00:00+05:30  
3          238 2025-09-02            4755 2012-08-25 03:00:00+05:30  
4          238 2025-09-02            4755 2012-08-25 04:00:00+05:30

---

- Use the column asia_timezone and convert it into the UTC timezone. 
- Use the function `tz_convert` to convert the timezone.

---

In [127]:
# change the asia time zone to UTC
data['utc_timezone'] = data.asia_timezone.dt.tz_convert('UTC')

In [128]:
data.head()

ID            Datetime  Count  month month_name  day_name  day_of_week  \
0   0 2012-08-25 00:00:00      8      8     August  Saturday            5   
1   1 2012-08-25 01:00:00      2      8     August  Saturday            5   
2   2 2012-08-25 02:00:00      6      8     August  Saturday            5   
3   3 2012-08-25 03:00:00      2      8     August  Saturday            5   
4   4 2012-08-25 04:00:00      2      8     August  Saturday            5   

   day_of_year      today  day_difference             asia_timezone  \
0          238 2025-09-02            4756 2012-08-25 00:00:00+05:30   
1          238 2025-09-02            4755 2012-08-25 01:00:00+05:30   
2          238 2025-09-02            4755 2012-08-25 02:00:00+05:30   
3          238 2025-09-02            4755 2012-08-25 03:00:00+05:30   
4          238 2025-09-02            4755 2012-08-25 04:00:00+05:30   

               utc_timezone  
0 2012-08-24 18:30:00+00:00  
1 2012-08-24 19:30:00+00:00  
2 2012-08-24 20:30:00+00:00  
3 2012-08-24 21:30:00+00:00  
4 2012-08-24 22:30:00+00:00

In [129]:
data[['asia_timezone', 'utc_timezone']]

asia_timezone              utc_timezone
0     2012-08-25 00:00:00+05:30 2012-08-24 18:30:00+00:00
1     2012-08-25 01:00:00+05:30 2012-08-24 19:30:00+00:00
2     2012-08-25 02:00:00+05:30 2012-08-24 20:30:00+00:00
3     2012-08-25 03:00:00+05:30 2012-08-24 21:30:00+00:00
4     2012-08-25 04:00:00+05:30 2012-08-24 22:30:00+00:00
...                         ...                       ...
18283 2014-09-25 19:00:00+05:30 2014-09-25 13:30:00+00:00
18284 2014-09-25 20:00:00+05:30 2014-09-25 14:30:00+00:00
18285 2014-09-25 21:00:00+05:30 2014-09-25 15:30:00+00:00
18286 2014-09-25 22:00:00+05:30 2014-09-25 16:30:00+00:00
18287 2014-09-25 23:00:00+05:30 2014-09-25 17:30:00+00:00

[18288 rows x 2 columns]

---

***Select a random date***

---

In [130]:
data['asia_timezone'][18287]

Timestamp('2014-09-25 23:00:00+0530', tz='Asia/Calcutta')

In [131]:
data['utc_timezone'][18287]

Timestamp('2014-09-25 17:30:00+0000', tz='UTC')

***You can see that time difference is 5 hours 30 minutes.***

---
---

#### `READING DATA WITH UNIX TIMESTAMP.`

- A UNIX timestamp is a way of storing a specific date and time. The timestamp is a ten digit number which represents the number of seconds that have passed since midnight on the 1st January 1970, UTC time.

---

In [133]:
# read data
data_with_unix_ts = pd.read_csv("C:\Arun\Data/data_with_timestamp.csv")

In [134]:
# view the data
data_with_unix_ts.head()

ID   timestamp  Count
0   0  1345852800      8
1   1  1345856400      2
2   2  1345860000      6
3   3  1345863600      2
4   4  1345867200      2

In [135]:
# convert the unix timestamp to datetime.
data_with_unix_ts.timestamp = pd.to_datetime(data_with_unix_ts.timestamp, unit='s')

In [136]:
# view the top rows
data_with_unix_ts.head()

ID           timestamp  Count
0   0 2012-08-25 00:00:00      8
1   1 2012-08-25 01:00:00      2
2   2 2012-08-25 02:00:00      6
3   3 2012-08-25 03:00:00      2
4   4 2012-08-25 04:00:00      2